databend-jdbc/src/main/java/com/databend/jdbc/DatabendConnection.java

```diff
@@ -984,4 +989,120 @@ void logout() throws SQLException {
+            heartbeatFuture = getScheduler().schedule(this, delay, MILLISECONDS);
+        }
+
+        private ArrayList<QueryLiveness> queryLiveness() {
+            ArrayList<QueryLiveness> arr = new ArrayList<>();
+            for (DatabendStatement stmt : statements.keySet()) {
+                QueryLiveness ql = stmt.queryLiveness();
+                if (ql != null && !ql.stopped && ServerVersions.supportHeartbeat(ql.serverVersion)) {
+                    arr.add(ql);
+                }
+            }
+            return arr;
+        }
+
+        private void doHeartbeat(ArrayList<QueryLiveness> queryLivenesses ) {
+            long now = System.currentTimeMillis();
+            lastHeartbeatStartTimeMillis = now;
+            Map<String, ArrayList<String>> nodeToQueryID = new HashMap();
+            Map<String, QueryLiveness> queries = new HashMap();
+
+            for (QueryLiveness ql: queryLivenesses) {
+                if (now - ql.lastRequestTime.get() >= ql.resultTimeoutSecs * 1000 / 2) {
+                    nodeToQueryID.computeIfAbsent(ql.nodeID, k -> new ArrayList()).add(ql.queryID);
+                    queries.put(ql.queryID, ql);
+                }
+            }
+            if (nodeToQueryID.isEmpty())
+                return;
+
+            ObjectMapper mapper = new ObjectMapper();
+            Map<String, Object> map = new HashMap<>();
+            map.put("node_to_queries", nodeToQueryID);
+
+            try {
+                String body = mapper.writeValueAsString(map);
+
+                body = generalRequest(HEARTBEAT_PATH, body);
+                JsonNode toRemove = mapper.readTree(body).get("queries_to_remove");
+                if (toRemove.isArray()) {
+                    for (JsonNode element : toRemove) {
+                        String queryId = element.asText();
+                        queries.get(queryId).stopped = true;
+                    }
+                }
+            } catch (JsonProcessingException e) {
+                logger.warning("fail to encode heartbeat body: " + e);
+            } catch (SQLException e) {
+                logger.warning("fail to send heartbeat: " + e);
+            } catch (IOException e) {
+                logger.warning("fail to send heartbeat: " + e);
+                throw new RuntimeException(e);
+            }
+        }
+
+        public void onStartQuery(Long timeoutSecs) {
+            synchronized (DatabendConnection.this) {
+                if (timeoutSecs * 1000 / 4 < heartbeatIntervalMillis) {
+                    heartbeatIntervalMillis = timeoutSecs * 1000 / 4;
+                    if (heartbeatFuture != null) {
+                        heartbeatFuture.cancel(false);
+                        heartbeatFuture = null;
+                    }
+                }
+                if (heartbeatFuture == null)
+                    scheduleHeartbeat();
+            }
+        }
+
+        @Override
+        public void run() {
+            heartbeatFuture = null;
```

Comment:
`heartbeatFuture = null` is out of  `synchronized ` , may cause concurrency problems

Reply:
fixed, thank you